caliper: init add 2.8.0

vsoch opened the pull request:
Adding LLNL/Caliper along with one required dependency, LLNL/Adiak (v0.2.2).

This is the first day I've tried developing a Nix package, so I apologize in advance for any oversights. I would like to add Caliper, and along the way I discovered that adiak would be needed as well. For the time being I'm limited to installing on Linux (the most common install I believe) and I've added most variants that are currently supported in Nixpkgs! I held off adding variants for niche software that is only needed by a very small set of specialized users.

For context, I work at a national lab, and am looking for streamlined ways to build developer environments in containers! I've been very impressed with Nix (the nixery! devbox that uses it!) and I'm hoping contributing packages that are typically for the HPC community can be relatively easy and something I can continue doing. I started this morning knowing absolutely nothing, and this really wasn't so bad! I wish I had tried a little earlier in my daily strugglebus with installing some of our software. :laughing: 

Thank you!

Signed-off-by: Vanessasaurus <814322+vsoch@users.noreply.github.com>

###### Description of changes

Addition of Caliper:
- https://github.com/llNL/caliper: A library used by other tools for performance metrics

Addition of Adiak:
- https://github.com/llNL/adiak: Also a library, the metadata portion of caliper above!


###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin

(For the above, I'm currently just starting with Linux support for containerization, I don't anticipate anyone using other OS will want to use nix, at least for the time being!)

I also want to ask about testing since it says "as applicable" - I was able to do a build for each of the packages - is that sufficient and is it automatically done here? What is the expectation for tests? If they are required, qhat is a good example of a minimal test? There are a lot of options, and I'd definitely appreciate guidance for this particular case. Thank you!

- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage) - I tried running this but it didn't work in GitHub codespaces, not sure if others have hit this?
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`) (no binary files, these are libs)

Should a release node be added for these, or is that just NixOS changes?

vsoch committed:
caliper: init add 2.8.0

Adding LLNL/Caliper along with one required dependency,
LLNL/Adiak (v0.2.2).

Signed-off-by: Vanessasaurus <814322+vsoch@users.noreply.github.com>

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+  # args.append("-DSOS_PREFIX=%s" % spec["sosflow"].prefix)
+
+  cmakeFlags = [
+       "-DENABLE_MPI=${onOffBool withMPI}"
+       "-DBUILD_SHARED_LIBS=${onOffBool shared}"
+       "-DWITH_LIBDW=${onOffBool elfutilsSupport}"
```
```suggestion
     "-DENABLE_MPI=${onOffBool withMPI}"
     "-DBUILD_SHARED_LIBS=${onOffBool shared}"
     "-DWITH_LIBDW=${onOffBool elfutilsSupport}"
```
and so on

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+  # variant("sosflow", default=False, description="Enable SOSflow support")
+  # variant("fortran", default=False, description="Enable Fortran support")
+
+  buildInputs = [ cmake python3 papi adiak ] ++
+    lib.optional libpfmSupport libpfm ++
+    lib.optional libunwindSupport libunwind ++
+    lib.optional elfutilsSupport elfutils ++
+
+    # These shouldn't be both provided
+    lib.optional withOpenmpi openmpi ++
+    lib.optional withMpich mpich;
```
```suggestion
  buildInputs = [ cmake python3 papi adiak ]
    ++ lib.optional libpfmSupport libpfm
    ++ lib.optional libunwindSupport libunwind
    ++ lib.optional elfutilsSupport elfutils
    ++ lib.optional withOpenmpi openmpi
    ++ lib.optional withMpich mpich;
```
this should be asserted

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+    platforms = platforms.linux;
+  };
+
+  # This means provided natively in Linux I think?
+  nativeBuildInputs = [ ];
+
```
```suggestion
```

setting nativeBuildInputs to an empty list does nothing.

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+    lib.optional withOpenmpi openmpi ++
+    lib.optional withMpich mpich;
+
+  # TODO how do these translate over (and is it needed)?
```
You probably want to set them to paths in the nix store. If they are required I can't tell you.

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+    sha256 = "17807b364b5ac4b05997ead41bd173e773f9a26ff573ff2fe61e0e70eab496e4";
+  };
+
+  meta = with lib; {
```
meta always must be last.

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+  pname = "caliper";
+  version = "2.8.0";
+
+  src = fetchurl {
```
Use fetchFromGitHub

vsoch replied:
I wasn't sure how this translates, and I don't have my nix setup anymore to test it out, so left as is for now (that I know was working)

SuperSandro2000 replied:
If you don't have nix anymore, how do you plan to maintain this package?

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+   withMPI = (withMpich == true || withOpenmpi == true);
+in
+
+# We can't have both mpis being used (is this the right way to do this?)
+assert (withMpich && !withOpenmpi) || (!withMpich && withOpenmpi) || (!withMpich && !withOpenmpi);
```
```suggestion
assert (withMpich && !withOpenmpi) || (!withMpich && withOpenmpi) || (!withMpich && !withOpenmpi);
```

Please convert that into a throw with a message which only triggers if both are true.

vsoch replied:
I didn't see how the suggested change above was different?

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,96 @@
+{ lib, stdenv, fetchurl, python3, cmake, adiak, papi
+, withCuda ? false, cudatoolkit
+, withRocm ? false, rocm-runtime
+, withFortran? false, gfortran
+, withOpenmpi? false, openmpi
+, withMpich ? false, mpich
+, libunwindSupport ? true, libunwind
+, libpfmSupport ? true, libpfm
+, elfutilsSupport ? true, elfutils
```
Do we really need options for all of them?

```suggestion
, withCuda ? false, cudatoolkit
, withRocm ? false, rocm-runtime
, withFortran ? false, gfortran
, withOpenmpi ? false, openmpi
, withMpich ? false, mpich
, libunwindSupport ? true, libunwind
, libpfmSupport ? true, libpfm
, elfutilsSupport ? true, elfutils
```

SuperSandro2000 commented on pkgs/development/libraries/adiak/default.nix:
```diff
@@ -0,0 +1,57 @@
+    fetchSubmodules = true;
+  };
+
+  meta = with lib; {
```
meta goes last in the attrset

SuperSandro2000 replied:
yes

SuperSandro2000 commented on pkgs/development/libraries/adiak/default.nix:
```diff
@@ -0,0 +1,57 @@
+    platforms = platforms.linux;
+  };
+
+  buildInputs = [ cmake ] ++
```
cmake must be in nativeBuildInputs

SuperSandro2000 commented on pkgs/development/libraries/adiak/default.nix:
```diff
@@ -0,0 +1,57 @@
+  # TODO how do these translate over (and is it needed)?
+  # args.append("-DMPI_CXX_COMPILER=%s" % self.spec["mpi"].mpicxx)
+  # args.append("-DMPI_C_COMPILER=%s" % self.spec["mpi"].mpicc)
+  cmakeFlags = [
+       "-DENABLE_MPI=${onOffBool withMPI}"
+       "-DBUILD_SHARED_LIBS=${onOffBool shared}"
+       "-DENABLE_TESTS=OFF"
+  ];
```
```suggestion
  cmakeFlags = [
     "-DENABLE_MPI=${onOffBool withMPI}"
     "-DBUILD_SHARED_LIBS=${onOffBool shared}"
     "-DENABLE_TESTS=OFF"
  ];
```

SuperSandro2000 reviewed (changes requested):
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

luzpaz commented:
Is there anything left to do here ?

vsoch commented:
@luzpaz contributing a package was so hard I think I largely gave up.

SuperSandro2000 commented on pkgs/development/libraries/adiak/default.nix:
```diff
@@ -0,0 +1,57 @@
+
+let
+   onOffBool = b: if b then "ON" else "OFF";
+   withMPI = (withMpich == true || withOpenmpi == true);
```
```suggestion
   withMPI = withMpich == true || withOpenmpi == true;
```

SuperSandro2000 reviewed (changes requested):
The commit title should be ``init *at* 2.8.0`` and the review commits should be squashed.

SuperSandro2000 commented:
> Is there anything left to do here ?

Yes, some review comments are left unresolved but if the author prefers we can close the PR instead.

vsoch commented:
Okay, rebased and squashed, and addressed the review comments that I knew how to! I opened this 14 months ago, and since then don't have my same environment (so I don't have a way to test) but I don't want to just give up on the contribution. I had really big hopes for using Nix but it feels like it's (overall) too hard.

Janik-Haag commented on maintainers/maintainer-list.nix:
```diff
@@ -19554,6 +19554,12 @@
     githubId = 7677567;
     name = "Victor SENE";
   };
+  vsoch = {
+    email = "contact@vsoch.com";
+    github = "vsoch";
+    githubId = 814322;
+    name = "Vanessa Sochat";
+  };
```
This should be it's own commit called `maintainers: add vsoch`

vsoch replied:
Done.

vsoch committed:
maintainers: add vsoch

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

SuperSandro2000 commented:
> The commit title should be `init *at* 2.8.0` and the review commits should be squashed.

This was not fixed

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,92 @@
+     "-DENABLE_MPI=${onOffBool withMPI}"
+     "-DBUILD_SHARED_LIBS=${onOffBool shared}"
+     "-DWITH_LIBDW=${onOffBool elfutilsSupport}"
+       "-DWITH_LIBPFM=${onOffBool libpfmSupport}"
+       "-DWITH_FORTRAN=${onOffBool withFortran}"
+       "-DWITH_CUPTI=${onOffBool withCuda}"
+       "-DWITH_NVTX=${onOffBool withCuda}"
+       "-DWITH_ROCTRACER=${onOffBool withRocm}"
+       "-DWITH_LIBUNWIND=${onOffBool libunwindSupport}"
+       "-DWITH_ROCTX=${onOffBool withRocm}"
+       "-DWITH_ADIAK=ON"
+       "-DWITH_PAPI=ON"
+       "-DWITH_SAMPLER=ON"
+       "-DWITH_GOTCHA=OFF"
+       "-DWITH_SOSFLOW=OFF"
+       "-DBUILD_TESTING=OFF"
+       "-DBUILD_DOCS=OFF"
```
```suggestion
     "-DWITH_LIBPFM=${onOffBool libpfmSupport}"
     "-DWITH_FORTRAN=${onOffBool withFortran}"
     "-DWITH_CUPTI=${onOffBool withCuda}"
     "-DWITH_NVTX=${onOffBool withCuda}"
     "-DWITH_ROCTRACER=${onOffBool withRocm}"
     "-DWITH_LIBUNWIND=${onOffBool libunwindSupport}"
     "-DWITH_ROCTX=${onOffBool withRocm}"
     "-DWITH_ADIAK=ON"
     "-DWITH_PAPI=ON"
     "-DWITH_SAMPLER=ON"
     "-DWITH_GOTCHA=OFF"
     "-DWITH_SOSFLOW=OFF"
     "-DBUILD_TESTING=OFF"
     "-DBUILD_DOCS=OFF"
```
that formatting is kinda off

SuperSandro2000 commented on pkgs/development/libraries/caliper/default.nix:
```diff
@@ -0,0 +1,92 @@
+    ++ lib.optional withOpenmpi openmpi
+    ++ lib.optional withMpich mpich;
+
+  # TODO how do these translate over (and is it needed)?
```
I am very confused by this comment

vsoch commented:
I’m not interested to contribute, thanks.
